File: python/day11.py
def adjacent_locations(x, y, width, height):
    for dx in range(-1, 2):
        for dy in range(-1, 2):
            if dx == 0 and dy == 0:
                continue

            (nx, ny) = (x + dx, y + dy)

            if nx >= 0 and ny >= 0 and nx < width and ny < height:
                yield (nx, ny)

def solve():
    with open("problems/problem11") as f:
        lines = f.readlines()

    grid = [[int(x) for x in line.strip()] for line in lines if line.strip()]

    width = len(grid[0])
    height = len(grid)

    def increase(x, y):
        grid[y][x] += 1

        # Only flash once
        if grid[y][x] == 10:
            ret = 1

            for nx, ny in adjacent_locations(x, y, width, height):
                ret += increase(nx, ny)

            return ret
        else:
            return 0

    prob1 = 0
    prob2 = None

    for step in range(2**32):
        flashed_this_step = 0

        for x in range(width):
            for y in range(height):
                flashed_this_step += increase(x, y)

        for x in range(width):
            for y in range(height):
                if grid[y][x] > 9:
                    grid[y][x] = 0

        if step < 100:
            prob1 += flashed_this_step

        if not prob2 and flashed_this_step == width * height:
            prob2 = step + 1

        if step >= 100 and prob2:
            break

    print("Problem 1:", prob1)
    print("Problem 2:", prob2)


if __name__ == '__main__':
    solve()

File: src/day11.rs
use crate::get_input_lines;

const INPUT: &str = "problems/problem11";

fn checked_add(a: usize, b: isize) -> Option<usize> {
    if b >= 0 {
        a.checked_add(b as usize)
    } else {
        a.checked_sub(-b as usize)
    }
}

pub fn solve() -> crate::Result<()> {
    let lines = get_input_lines(INPUT)?;

    let mut grid = lines
        .map(|line| {
            line.chars()
                .map(|x| -> Option<u8> { Some(x.to_digit(10)? as u8) })
                .collect::<Option<Vec<_>>>()
        })
        .collect::<Option<Vec<_>>>()
        .ok_or(crate::Error::InvalidInput)?;

    let width = grid.first().ok_or(crate::Error::NoInput)?.len();

    let height = grid.len();

    pub fn increase(
        x: usize,
        y: usize,
        grid: &mut Vec<Vec<u8>>,
        width: usize,
        height: usize,
    ) -> u64 {
        grid[y][x] += 1;

        // Only flash once
        if grid[y][x] == 10 {
            let mut ret = 1;

            for dx in -1..=1 {
                for dy in -1..=1 {
                    if dx == 0 && dy == 0 {
                        continue;
                    }

                    if let (Some(nx), Some(ny)) = (checked_add(x, dx), checked_add(y, dy)) {
                        if nx >= width || ny >= height {
                            continue;
                        }

                        ret += increase(nx, ny, grid, width, height);
                    }
                }
            }

            ret
        } else {
            0
        }
    }

    let mut prob1 = 0;
    let mut prob2 = None;

    for step in 0.. {
        let mut flashed_this_step = 0;

        for y in 0..height {
            for x in 0..width {
                flashed_this_step += increase(x, y, &mut grid, width, height);
            }
        }

        for y in 0..height {
            for x in 0..width {
                if grid[y][x] > 9 {
                    grid[y][x] = 0;
                }
            }
        }

        if step < 100 {
            prob1 += flashed_this_step;
        }

        if prob2 == None && flashed_this_step == (width * height) as u64 {
            prob2 = Some(step + 1);
        }

        if step >= 100 && prob2.is_some() {
            break;
        }
    }

    println!("Problem 1: {}", prob1);
    println!("Problem 2: {}", prob2.unwrap());

    Ok(())
}
